src/src/PluginStructs/P164_data_struct.cpp

```diff
@@ -0,0 +1,889 @@
+#define IS_ENS160_SEQ_ACK_COMPLETE(x)       (ENS160_SEQ_ACK_COMPLETE == (ENS160_SEQ_ACK_COMPLETE & (x)))
+
+// ENS160 STATUS bitfields
+#define ENS160_STATUS_STATAS        0x80    // STATAS: Indicates that an OPMODE is rumming
+#define ENS160_STATUS_STATER        0x40    // STATER: High indicated that an error is detected
+#define ENS160_STATUS_VALIDITY      0x0C    // VALIDITY FLAG
+#define ENS160_STATUS_NEWDAT        0x02    // NEWDAT: 1= New data in data registers available
+#define ENS160_STATUS_NEWGPR        0x01    // NEWGRP: 1= New data in GRP_READ registers available 
+
+// Checkers for bitfields in STATUS register
+#define IS_NEWDAT(x)                (ENS160_STATUS_NEWDAT == (ENS160_STATUS_NEWDAT & (x)))
+#define IS_NEWGPR(x)                (ENS160_STATUS_NEWGPR == (ENS160_STATUS_NEWGPR & (x)))
+#define IS_NEW_DATA_AVAILABLE(x)    (0 != ((ENS160_STATUS_NEWDAT | ENS160_STATUS_NEWGPR ) & (x)))
+#define GET_STATUS_VALIDITY(x)      (((x) & ENS160_STATUS_VALIDITY) >> 2)
+
+// TODO: add comment on this
+#define CONVERT_RS_RAW2OHMS_I(x)    (1 << ((x) >> 11))
+#define CONVERT_RS_RAW2OHMS_F(x)    (pow (2, (float)(x) / 2048))
+
+// Form IDs used on the device setup page. Should be a short unique string.
+#define P164_GUID_TEMP_T     "f08"
+#define P164_GUID_TEMP_V     "f09"
+#define P164_GUID_HUM_T      "f10"
+#define P164_GUID_HUM_V      "f11"
+
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Constructor                                                                                   //
+///////////////////////////////////////////////////////////////////////////////////////////////////
+P164_data_struct::P164_data_struct(struct EventStruct *event) :
+  i2cAddress(P164_I2C_ADDR)
+{
+  #ifdef P164_ENS160_DEBUG
+    Serial.println(F("ENS160: Constructor"));
+    Serial.print("ENS160: I2C address =");
+    Serial.println(i2cAddress, HEX);
+  #endif
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Initialization of the connected device                                                        //
+///////////////////////////////////////////////////////////////////////////////////////////////////
+bool P164_data_struct::begin()
+{
+  if (!start(i2cAddress)) {
+    Serial.println(F("P164: device initialization ***FAILED***"));
+    return false;
+  }
+  setMode(ENS160_OPMODE_STD);
+  #ifdef P164_ENS160_DEBUG
+    Serial.println(F("P164: begin(): success"));
+  #endif
+  return true;
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Fetch the processed device values as stored in the software object                            //
+///////////////////////////////////////////////////////////////////////////////////////////////////
+bool P164_data_struct::read(float& tvoc, float& eco2)
+{
+  bool success = measure();
+  tvoc = (float)_data_tvoc;
+  eco2 = (float)_data_eco2;
```

Comment:
Don't use C-style casts. Also when assigning to a float, you don't need to cast it.